stream: hide afterWriteTickInfo property

Hakerh400 opened the pull request:
Pull request https://github.com/nodejs/node/pull/30710 has introduced `afterWriteTickInfo` property for optimizing synchronous write completions and it is backported to v12.x as a semver-minor release, but it breaks use case of JSON stringifying writable stream object. This PR hides that property behind a symbol.

Fixes https://github.com/nodejs/node/issues/31277

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

BridgeAR commented:
@Hakerh400 maybe also add a regression test?

mscdex commented:
Perhaps we can find a better way to do this?

```
                                                  confidence improvement accuracy (*)   (**)  (***)
  streams/creation.js kind='duplex' n=50000000           ***    -51.76 %       ±0.96% ±1.27% ±1.66%
  streams/creation.js kind='transform' n=50000000        ***    -48.81 %       ±2.01% ±2.70% ±3.55%
  streams/creation.js kind='writable' n=50000000         ***    -64.69 %       ±2.57% ±3.47% ±4.59%
```

BridgeAR commented:
Using a symbol instead should have less performance impact.

mscdex commented:
This also needs a test.

jasnell commented:
This is also likely semver-major.

Hakerh400 commented:
Changed to a symbol and added a test.

I don't think this should be a semver-major. The PR that introduced this property was released in a semver-minor release (and could propably be a semver-patch). This PR can be considered as a fix for a broken use case.

mscdex commented:
Benchmarks with symbol:

```
                                                  confidence improvement accuracy (*)   (**)  (***)
  streams/creation.js kind='duplex' n=50000000                   -1.51 %       ±2.24% ±2.99% ±3.93%
  streams/creation.js kind='transform' n=50000000          *      3.69 %       ±3.38% ±4.51% ±5.88%
  streams/creation.js kind='writable' n=50000000                  2.66 %       ±3.54% ±4.71% ±6.13%
```

Trott commented:
@jasnell LGTY?

ronag commented on lib/_stream_writable.js:
```diff
@@ -60,6 +60,8 @@ const {
 
 const { errorOrDestroy } = destroyImpl;
 
+const afterWriteTickInfo = Symbol('afterWriteTickInfo');
```
I think we prefix symbols with `k` i.e. `kAfterWriteTickInfo`

Hakerh400 replied:
Done

Hakerh400 commented:
> I'm a little curious about this stringify the stream state use case. That's not really something I believe we support and expecting stringify/parse to work could be a bit dangerous, e.g. if you have callbacks buffered.

I agree, it is not documented anywhere that a stream object should not contain circular structures. That is why initially I didn't add any test, but @BridgeAR and @mscdex were explicit about adding a test. The point of this PR is to fix a use case in which a code stopped working after a semver-minor release, which may leave users unprepared. I am more than happy to remove the test if we all agree.

ronag commented:
The test is fine. I’m actually asking whether we should add further tests.

jasnell commented on lib/_stream_writable.js:
```diff
@@ -148,7 +150,7 @@ function WritableState(options, stream, isDuplex) {
   // Storage for data passed to the afterWrite() callback in case of
   // synchronous _write() completion.
-  this.afterWriteTickInfo = null;
+  this[kAfterWriteTickInfo] = null;
```
Sorry, I should have been clearer. This change is good, but the `afterWriteTickInfo` property should still be retained as a non-enumerable alias.

e.g.

```js
Object.defineProperty(this, 'afterWriteTickInfo', {
  enumerable: false,
  get() { return this[kAfterWriteTickInfo]; },
  set(val) { this[kAfterWriteTickInfo] = val; }
});
```

The alias can then be separated deprecated.

mscdex replied:
I believe `Object.defineProperty()` is what was used originally and calling that at every invocation is what caused the significant performance regression. Might be able to add it to `WritableState`'s prototype once on initial module load though.

mscdex replied:
Actually, this property has only existed for two months (since 2205f85b2caadee425a0a86bd8cc3bcb889e4bfe). I doubt anyone has written any code that relies on this internal state property, so an alias probably isn't even necessary...

Hakerh400 replied:
@jasnell  What is the purpose of defining a symbol property and then adding an extra allias for it that accesses the original property through a getter/setter? Wouldn't it destroy performance and add redundant complexity?

BTW, I'm not sure if this was discussed before, but... why don't we switch to private class fields. They are much faster than public attributes and they are fully hidden from user.

jasnell replied:
@mscdex good points! if the property has gone out in a release however, removing it still makes it semver-major unfortunately. Adding the alias to the prototype would work.

@Hakerh400 ... using the symbol avoids the performance hit, adding the alias avoids it needing to be a semver-major change.

Hakerh400 replied:
Ok, added the alias

mscdex commented on lib/_stream_writable.js:
```diff
@@ -148,7 +150,12 @@ function WritableState(options, stream, isDuplex) {
   // synchronous _write() completion.
-  this.afterWriteTickInfo = null;
+  this[kAfterWriteTickInfo] = null;
+  ObjectDefineProperty(this, 'afterWriteTickInfo', {
```
This should go on the prototype and be moved to the parent scope so that it only runs once.

Hakerh400 replied:
Done.

Hakerh400 commented:
CI is green.

@jasnell @ronag Any objections left?

lundibundi commented on lib/_stream_writable.js:
```diff
@@ -185,6 +187,12 @@ function WritableState(options, stream, isDuplex) {
 }
 
+ObjectDefineProperty(WritableState.prototype, 'afterWriteTickInfo', {
+  enumerable: false,
```
Nit: this is the default and can be omitted if you want.

addaleax replied:
I’d prefer explicit over implicit.

lundibundi replied:
I don't have a strong opinion and fine either way.
I just remembered this PR https://github.com/nodejs/node/pull/31187 that actually removes explicit checks and thought we were moving in this direction and wanted to point this out. Though now, after checking the `stream_readable` and noticing that it still has all of the `enumerable: false` properties it doesn't look like it and the aforementioned PR is actually an inconsistency here :smile: .

Hakerh400 replied:
@lundibundi Thanks for the suggestion. Object property descriptor attributes are `false` by default because `undefined` is falsy when coerced to boolean. JavaScript engine searches accross the prototype chain and the first defined value (or `undefined` if the last null-prototype object has no such property defined) is casted to a boolean, and the search takes some time.

Performance is not critical here though (because this code runs only once), but in my understanding, omitting these properties is generally an anti-pattern.

antsmartian replied:
So are we deciding that we need an explicit `enumerable: false`? We had an earlier discussion that we can omit them. Since I was refactoring to group all these define property into a single block, writing up enumerable false for each property wasn't something that I had prefer. Thoughts?

ronag replied:
I find having the property explicitly defined can be confusing for some developers (it was for me in the beginning), i.e. it seem to mean that there is something special going on rather than being just a regular getter/setter.

No strong opinion though...

Hakerh400 replied:
Do we have a consensus here? I am not opposed to make it implicit if the PR cannot land otherwise, but after reading https://github.com/nodejs/node/pull/31187#discussion_r370340253 it sounds like the winds are blowing the opposite direction.

Note: This commit should also be backported to v12.x to actually fix https://github.com/nodejs/node/issues/31277.

lundibundi commented on test/parallel/test-json-stringify-stdout.js:
```diff
@@ -0,0 +1,9 @@
+'use strict';
+
+// Make sure stdout does not contain circular structures
```
Nit:
```suggestion
// Make sure stdout does not contain circular structures.
```

Hakerh400 commented:
Given that https://github.com/nodejs/node/pull/31187 has landed without explicit enumerability, I tend to assume that we have decided what the consensus regarding this topic is.

Removed default `enumerable` property. PTAL.

Hakerh400 commented:
> I’m okay with landing this with the understanding that a semver-major revert PR is opened afterwards.

Seems reasonable. I'm ready to revert this in a separate PR once it lands.
